packages/@react-aria/autocomplete/src/useAutocomplete.ts

```diff
@@ -0,0 +1,242 @@
+  /** Props for the collection, to be passed to collection's respective aria hook (e.g. useMenu). */
+  collectionProps: CollectionOptions,
+  /** Ref to attach to the wrapped collection. */
+  collectionRef: RefObject<HTMLElement | null>,
+  /** A filter function that returns if the provided collection node should be filtered out of the collection. */
+  filterFn: (nodeTextValue: string) => boolean
+}
+
+/**
+ * Provides the behavior and accessibility implementation for a autocomplete component.
+ * A autocomplete combines a text input with a collection, allowing users to filter the collection's contents match a query.
+ * @param props - Props for the autocomplete.
+ * @param state - State for the autocomplete, as returned by `useAutocompleteState`.
+ */
+export function useAutocomplete(props: AriaAutocompleteOptions, state: AutocompleteState): AutocompleteAria {
+  let {
+    collectionRef,
+    defaultFilter
+  } = props;
+
+  let collectionId = useId();
+  let timeout = useRef<ReturnType<typeof setTimeout> | undefined>(undefined);
+  let delayNextActiveDescendant = useRef(false);
+  let lastCollectionNode = useRef<HTMLElement>(null);
+
+  let updateActiveDescendant = useCallback((e) => {
+    let {detail} = e;
+    clearTimeout(timeout.current);
+    e.stopPropagation();
+    if (detail?.id != null) {
+      if (delayNextActiveDescendant.current) {
+        timeout.current = setTimeout(() => {
+          state.setFocusedNodeId(detail.id);
+        }, 500);
+      } else {
+        state.setFocusedNodeId(detail.id);
+      }
+    } else {
+      state.setFocusedNodeId(null);
+    }
+
+    delayNextActiveDescendant.current = false;
+  }, [state]);
+
+  let callbackRef = useEffectEvent((collectionNode: HTMLElement | null) => {
+    // When typing forward, we want to delay the setting of active descendant to not interrupt the native screen reader announcement
+    // of the letter you just typed. If we recieve another UPDATE_ACTIVEDESCENDANT call then we clear the queued update
+    // We track lastCollectionNode to do proper cleanup since callbackRefs just pass null when unmounting. This also handles
+    // React 19's extra call of the callback ref in strict mode
+    if (collectionNode != null) {
+      lastCollectionNode.current?.removeEventListener(UPDATE_ACTIVEDESCENDANT, updateActiveDescendant);
+      lastCollectionNode.current = collectionNode;
+      collectionNode.addEventListener(UPDATE_ACTIVEDESCENDANT, updateActiveDescendant);
+    } else {
+      lastCollectionNode.current?.removeEventListener(UPDATE_ACTIVEDESCENDANT, updateActiveDescendant);
+    }
+  });
+  // Make sure to memo so that React doesn't keep registering a new event listeners on every rerender of the wrapped collection,
+  // especially since callback refs's node parameter is null when they cleanup so we can't even clean up properly
+  let mergedCollectionRef = useObjectRef(useMemo(() => mergeRefs(collectionRef, callbackRef), [collectionRef, callbackRef]));
```

Comment:
Something since e3c3c35d269ab52c830993cb6f75808f0ab5aa1a broke the registering of the event listeners on Listbox. Something with the memo either here or in Listbox when merging refs but for some reason works just fine for Menu...